atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,208 @@
+        return self.koji_build_id or nvr_build_id
+
+    def assemble_srpm_url(self, base_url, srpm_filename, sign_key=None):
+        """Assemble the URL used to fetch an SRPM file
+
+        :param base_url: str, Koji root base URL with the given build artifacts
+        :param srpm_filename: str, name of the SRPM file
+        :param sign_key: str, key used to sign the SRPM, as listed in the signing intent
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        url_components = [base_url, 'src', srpm_filename]
+        if sign_key:
+            url_components[1:1] = ['data', 'signed', sign_key]
+        return '/'.join(url_components)
+
+    def get_srpm_urls(self, build_id, sigkeys=None, insecure=False):
+        """Fetch SRPM download URLs for each image generated by a build
+
+        Build each possible SRPM URL and check if the URL is available,
+        respecting the signing intent preference order.
+
+        :param build_id: int, Koji build id of the container image we want SRPMs for
+        :param sigkeys: list, strings for keys which signed the srpms to be fetched
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        srpm_urls = set()
+        archives = self.session.listArchives(build_id, type='image')
+        if not sigkeys:
+            sigkeys = ['']
+
+        for archive in archives:
+            rpms = self.session.listRPMs(imageID=archive['id'])
+            for rpm in rpms:
+                rpm_hdr = self.session.getRPMHeaders(rpm['id'], headers=['SOURCERPM'])
+                srpm_filename = rpm_hdr['SOURCERPM']
+                rpm_build = self.session.getBuild(rpm['build_id'])
+                path_info = get_koji_path_info(self.workflow, self.koji_fallback)
+                base_url = path_info.build(rpm_build)
+                for sigkey in sigkeys:
+                    url_candidate = self.assemble_srpm_url(base_url, srpm_filename, sigkey)
+                    req_session = get_retrying_requests_session()
+                    request = req_session.head(url_candidate, verify=not insecure)
+                    if request.status_code == requests.codes.ok:
+                        srpm_url = url_candidate
+                        break
+                try:
+                    srpm_urls.add(srpm_url)
+                except NameError:
+                    raise RuntimeError('Could not satisfy signing intent for {}'.format(rpm['nvr']))
+        return list(srpm_urls)
+
+    def get_signing_intent(self, build_id):
+        """Get the signing intent to be used to fetch files from Koji
+
+        :param build_id: int, Koji build id of the container image which owns the signing intent
+        :return: dict, signing intent object as per atomic_reactor/schemas/config.json
+        """
+        odcs_config = get_config(self.workflow).get_odcs_config()
+        if not self.signing_intent:
+            image_build = self.session.getBuild(build_id)
```

Comment:
Can we avoid this second call to `getBuild()` by caching the result of the earlier call (in `get_image_build_id()`)?

Reply:
The first call to getBuild is only necessary if a koji_nvr is provided. If a signing intent is provided, the second call is not necessary. I added a new `self.koji_build` to store the build in case the first call is needed.